walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -11,13 +11,64 @@
 // WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 // See the License for the specific language governing permissions and
 // limitations under the License.
-
 package com.google.sps;
 
-import java.util.Collection;
-
-public final class FindMeetingQuery {
+import java.util.*;
+ 
+public final class FindMeetingQuery { 
+  private ArrayList<TimeRange> viableTimeSlots = new ArrayList<TimeRange>(); 
+ 
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    ArrayList<TimeRange> busyTimes = new ArrayList<TimeRange>();
+    ArrayList<TimeRange> allBusyTimes = new ArrayList<TimeRange>();
+    for(Event e : events) {
+    if(eventContains(e.getAttendees(),request.getAttendees())) { // Ensure event includes requireds
+      busyTimes.add(e.getWhen());  // Build list of only requireds
+      allBusyTimes.add(e.getWhen()); // Add required list to overall list
+    }  
+      if(eventContains(e.getAttendees(), request.getOptionalAttendees()))
+        allBusyTimes.add(e.getWhen()); // Add optionals to overall list
+    }
+    
+    // Generate available times including optionals
+    generatePotentialTimes(allBusyTimes, (int)request.getDuration());
+ 	if(viableTimeSlots.size() == 0) {
+    // Generate available times with only requireds.
+    generatePotentialTimes(busyTimes, (int)request.getDuration());
+    }
+    return viableTimeSlots;
+  }
+ 
+  private boolean eventContains(Collection<String> eventAttendees, Collection<String> targetAttendees) {
+    for(String attendee : targetAttendees) {
+        if(eventAttendees.contains(attendee))
+          return true;
+    }
+    return false;
+  }
+ 
+  private void generatePotentialTimes(ArrayList<TimeRange> busyTimes, int duration) {
+    // Ensure list is in chronological order
+    Collections.sort(busyTimes, TimeRange.ORDER_BY_START); 
+    int startTime = TimeRange.START_OF_DAY;
+    // Find gaps between busy times
+    for(TimeRange timeRange : busyTimes) {
+      if(timeRange.start() >= startTime) { // Check that this busy time starts after previous one ends
+        // Create gap between end of last busy time and this one
+        checkInsertTime(startTime, timeRange.start(), duration,false);   
+     }
+      // Set start of potential meeting time to end of this time slot.
+      // Max function helps resolve nested events.
```

Comment:
Remove this comment, and you can explain a little better in line 60 what the max will do since in line 60 you only explained one scenario.